sequencing-server/src/app.ts

```diff
@@ -234,4 +243,18 @@ app.use((err: any, _: Request, res: Response, next: NextFunction) => {
 
 app.listen(PORT, () => {
   logger.info(`connected to port ${PORT}`);
+  logger.info(`Worker pool initialized:
+              Total workers started: ${piscina.threads.length},
+              Heap Size per Worker: ${getEnv().SEQUENCING_MAX_WORKER_HEAP_MB} MB`);
 });
+
+if (getEnv().TRANSPILER_ENABLED === 'true') {
+  //log that the tranpiler is on
+  logger.info(`Background Transpiler is 'on'`);
+  // Immediately call the background transpiler
+  await backgroundTranspiler();
+  // Then continue calling it every 5 minutes
+  setInterval(async () => {
```

Comment:
discussed with @goetzrrGit - this will not wait for previous transpile to be done before setting timer - will update

Reply:
I think this might be better expressed as a combination of a `throttle` with a function that calls itself on completion. I think the current implementation could end up with a 10 minute gap if the background transpiler is still running when the next invocation is intended to occur rather than an immediate invocation.

in psuedocode
```typescript
const runBackgroundTranspiler = () => backgroundTranspiler.finally(() => throttle(runBackgroundTranspiler, 60 * 5 * 1000));

// initially kick it off
runBackgrounTranspiler();
```

Reply:
I did some testing and I think we are generating too much data on the heap with the throttle. I had the server up for about an hour in a meeting. 

```
[node] <--- Last few GCs --->
[node] 
[node] [1355:0x7f460ceb82a0]    82200 ms: Scavenge 2036.8 (2078.8) -> 2035.2 (2079.6) MB, 8.7 / 0.0 ms  (average mu = 0.152, current mu = 0.122) allocation failure; 
[node] [1355:0x7f460ceb82a0]    82219 ms: Scavenge 2037.4 (2079.6) -> 2036.2 (2081.1) MB, 16.8 / 0.0 ms  (average mu = 0.152, current mu = 0.122) allocation failure; 
[node] [1355:0x7f460ceb82a0]    82238 ms: Scavenge 2039.0 (2081.1) -> 2037.5 (2090.1) MB, 16.6 / 0.0 ms  (average mu = 0.152, current mu = 0.122) allocation failure; 
[node] 
[node] 
[node] <--- JS stacktrace --->
[node] 
[node] FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
[node] Program node --inspect=0.0.0.0 --experimental-vm-modules ./build/app.js exited with code null
```

Reply:
I did try playing around with your suggestion but I keep getting a heap out of bounds. 

```
<--- Last few GCs --->
[node] 
[node] [1655:0x7fad990f52a0]    97704 ms: Scavenge 2027.3 (2069.3) -> 2026.1 (2071.1) MB, 11.1 / 0.0 ms  (average mu = 0.183, current mu = 0.146) allocation failure; 
[node] [1655:0x7fad990f52a0]    97718 ms: Scavenge 2028.8 (2071.1) -> 2027.5 (2080.1) MB, 12.2 / 0.0 ms  (average mu = 0.183, current mu = 0.146) allocation failure; 
[node] [1655:0x7fad990f52a0]    97737 ms: Scavenge 2033.8 (2080.1) -> 2030.7 (2081.3) MB, 13.9 / 0.0 ms  (average mu = 0.183, current mu = 0.146) allocation failure; 
[node] 
[node] 
[node] <--- JS stacktrace --->
[node] 
[node] FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
[node] Program node --inspect=0.0.0.0 --experimental-vm-modules ./build/app.js exited with code null
```

Reply:
You have to have the recursive call inside the function that gets kicked off, so that latest invocation won't work. I'm surprised by the memory usage as I tested this with the following test code and had memory stabilize for me with <30MB heap and garbage collection seemed to be working.

Thoughts on why you might be seeing garbage collection not happening? What does the size of your cache look like over time?


```json
{
  rss: '111.5 MB -> Resident Set Size - total memory allocated for the process execution',
  heapTotal: '54.97 MB -> total size of the allocated heap',
  heapUsed: '27.52 MB -> actual memory used during the execution',
  external: '2.27 MB -> V8 external memory'
}
```

```typescript
import throttle from 'lodash.throttle';

async function someAsync() {
	await new Promise(resolve => setTimeout(resolve, 1));
	console.log(formatMemoryUsage(process.memoryUsage()));
	return;
}

const throttledSomeAsync = throttle(someAsync, 10);

function recursiveCall() {
	throttledSomeAsync().finally(recursiveCall)
}

recursiveCall()

function formatMemoryNumber(mem: number){
	return `${Math.round(mem / 1024 / 1024 * 100) / 100} MB`;
}

function formatMemoryUsage(memoryUsage: NodeJS.MemoryUsage) {
	return {
	  rss: `${formatMemoryNumber(memoryUsage.rss)} -> Resident Set Size - total memory allocated for the process execution`,
	  heapTotal: `${formatMemoryNumber(memoryUsage.heapTotal)} -> total size of the allocated heap`,
	  heapUsed: `${formatMemoryNumber(memoryUsage.heapUsed)} -> actual memory used during the execution`,
	  external: `${formatMemoryNumber(memoryUsage.external)} -> V8 external memory`,
	};
}
```

Reply:
AHA! I found it! Because anonymous functions `const functionName = () => { //body }` capture their context for execution, it causes memory to explode. Change it to the declared function like I have above and give it a try.

So
```typescript
// The throttle wrap is lightly stateful, so we need to hold onto the same instance.
const trottledBackgroundTranspiler = throttle(backgroundTranspiler, 60 * 5 * 1000);
// This will trigger another run on completion. We need this here so that we attach the finally call to the promise returned each time its called
function runBackgroundTranspiler() {
  return trottledBackgroundTranspiler().finally(runBackgroundTranspiler);
}

// initially kick it off
runBackgrounTranspiler();
```

Reply:
Hi @dyst5422, thank you for looking into this! The server is now stable after removing the arrow function and throttling at 10 milliseconds. I ran it for 10 minutes with periodic updates to the expansion logic, and the memory holds during my testing.

```
[node].  rss: '1251.54 MB -> Resident Set Size - total memory allocated for the process execution',
[node]   heapTotal: '80.97 MB -> total size of the allocated heap',
[node]   heapUsed: '44.73 MB -> actual memory used during the execution',
[node]   external: '3.68 MB -> V8 external memory'
```

Reply:
Doing further testing, I discovered that the HTTP GET requests to the server are not responding or returning data. Unfortunately, due to time constraints, I cannot investigate the cause of this issue as we're releasing a new version of Aerie tomorrow.